### PR TITLE
Feature: TLS for gRPC

### PR DESCRIPTION
**Pull request by sgammon**
This changeset fixes and closes cloudendpoints/esp#346, by adding the ability to invoke the gRPC channel constructor with SSL/TLS credentials based on nginx config. I am of course completely open to feedback both on my approach and implementation, especially because C++ is not my native language

Basically, a few config entries are added, which are interpreted at startup and then passed to the gRPC `SslCredentials` struct:

**New config file entries**:
- `grpc_roots`: Specify a file full of PEMs to use as authorities, when operating from the client's perspective (so, authorities for upstream gRPC backends).

- `grpc_client_key`: Kind of obvious. Lets you specify a PEM file to load and use as a client-side private key.

- `grpc_client_certificte`: Matches the `grpc_client_key` with a path to a PEM file containing a client-side certificate.

**Behavior:**
To enable TLS upstream to gRPC, simply prefix the `grpc_pass` address with `grpcs://`, like so:
```
location / {
  grpc_pass grpcs://127.0.0.1:1337 override;
}
```

In that same block, specify other directives to control how TLS works. Here is a full configuration:
```
location / {
  grpc_pass grpcs://127.0.0.1:1337 override;
  grpc_client_key /path/to/client/key.pem;
  grpc_client_certificate /path/to/client/cert.pem;
  grpc_roots /path/to/custom/roots.pem
}
```

If either `grpc_client_key` or `grpc_client_certificate` is specified, the other must be specified. Both are loaded and kept for when the gRPC channel is initialized.

If `grpc_roots` is 1) left unspecified, or 2) specified with the special value `default`, the default root authorities from gRPC are used.

Remaining items to do:
- [ ] Fail if a client cert, but not a key, is specified
- [ ] Probably some unit tests somewhere
- [ ] Testing in sandbox with our own services
- [ ] Eagerly validate the key and cert (don't wait until channel initialization to fail)

### CODE REVIEW

File: src/nginx/module.cc [hunk 1]
```diff
@@ -234,13 +234,61 @@ ngx_command_t ngx_esp_commands[] = {
...
+        // specified).
+        //
+        // Usage:
+        //   location / {
+        //     grpc_roots [/path/to/roots.pem|"default"];
+        //   }
+        //
+        ngx_string("grpc_roots"),
+        NGX_HTTP_LOC_CONF | NGX_CONF_TAKE1,
+        ConfigureGrpcAuthorities, NGX_HTTP_LOC_CONF_OFFSET, 0, nullptr,
```
Comment (sgammon) [inline src/nginx/module.cc]: documents and sets up the new `grpc_roots` configuration directive.

usage example:
```text
location / {
  grpc_roots /path/to/roots.pem;
}
```

or to use the gRPC defaults:
```text
location / {
  grpc_roots default;
}
```

File: src/nginx/module.cc [hunk 2]
```diff
@@ -234,13 +234,61 @@ ngx_command_t ngx_esp_commands[] = {
...
+        // also specify `grpc_client_certificate`.
+        //
+        // Usage:
+        //   location / {
+        //     grpc_client_key /path/to/key.pem;
+        //   }
+        //
+        ngx_string("grpc_client_key"),
+        NGX_HTTP_LOC_CONF | NGX_CONF_TAKE1,
+        ConfigureGrpcClientKey, NGX_HTTP_LOC_CONF_OFFSET, 0, nullptr,
```
Comment (sgammon) [inline src/nginx/module.cc]: defines the new `grpc_client_key` directive. used with the new `grpc_client_certificate` directive, like so:

```text
location / {
  grpc_client_key /path/to/key.pem;
  grpc_client_certificate /path/to/certificate.pem;
}
```

if one is defined, the other must be defined. both must contain PEM formatted crypto assets, which are passed to gRPC when creating a channel later on for the endpoint in `grpc_pass`.

File: src/nginx/util.h [hunk 3]
```diff
@@ -65,6 +65,11 @@ inline bool ngx_string_equal(const ngx_str_t &str1, const ngx_str_t &str2) {
   return str1.len == str2.len && !ngx_strncmp(str1.data, str2.data, str1.len);
 }
 
+// Check an ngx_str_t against a prefix ngx_str_t
+inline bool ngx_string_startswith(const ngx_str_t &subject, const ngx_str_t &prefix) {
+  return subject.len > prefix.len && ngx_strstr(subject.data, prefix.data) == 0;
```
Comment (sgammon) [inline src/nginx/util.h]: added a util method `ngx_string_startswith`. to opt-in to upstream TLS when communicating over gRPC, simply prefix the endpoint with `grpcs://`.

File: src/nginx/grpc.cc [hunk 4]
```diff
@@ -94,20 +97,85 @@ std::pair<Status, std::shared_ptr<::grpc::GenericStub>> GrpcGetStub(
...
+
   ::grpc::ChannelArguments channel_arguments;
 
   channel_arguments.SetMaxReceiveMessageSize(INT_MAX);
   channel_arguments.SetMaxSendMessageSize(INT_MAX);
 
-  auto result =
-      std::make_shared<::grpc::GenericStub>(::grpc::CreateCustomChannel(
-          address, ::grpc::InsecureChannelCredentials(), channel_arguments));
+  if (ngx_string_startswith(ngx_addr, kGrpcTlsPrefix)) {
```
Comment (sgammon) [inline src/nginx/grpc.cc]: if the `address` begins with `grpcs://`... then turn on upstream TLS, optionally with `grpc_roots` set (or not, in which case defaults are used). if a client key and cert are set, pass those in, too.

**Comment by qiwzhang** [comment]
Sorry for the late response.   I just updated your changes to the latest master and started the presumit tests.

You may have to format the code with clang-format 3.8.0 version as
clang-3.8.0/bin/clang-format -style=google -i

**Comment by qiwzhang** [comment]
As I expected, the presubmit test with for code format:

Checking coding styles.
Files not formatted: src/nginx/grpc.cc src/nginx/grpc.h src/nginx/module.cc src/nginx/util.h
Code style check failed.

**Comment by sgammon** [comment]
@qiwzhang, can you possibly tell me how you guys develop and debug with ESP? i'm specifically having trouble triggering the gRPC core (in C/C++) to emit debugging logs.

if you guys have procedures for this that would help immensely. i'm working on a less invasive approach, and i will make sure my patch is updated both with the new code, and that i check formatting before i file the PR again. thank you for your feedback

**Comment by qiwzhang** [comment]
We are using Nginx log,  you have to do two things to emit log
1) in Nginx config, make sure error_log is in debug level
2) build with debug symbol by "bazel build -c dbg"

**Pull request closed without merging**